### PR TITLE
[WIP] #430 make e2e csv full report all backends target

### PR DESCRIPTION
**Pull request by emabota**
This PR is aimed at the issue #430. 
It focus, initially, on the implementation of the e2e_test in go. 
I decided to start by aiming at a program that will work on this conditions: **ip_family=ipv4, backend=iptables, cluster_count=1**

Studying the current `test_e2e.sh` script I identified the main functions required to have a working `test_e2e.go` program. And documented it in [here](https://docs.google.com/document/d/1fanr95mp9Sc-wh9Ss0PFqu9yqUA6tdA2fRxUWq3lmmY/edit?usp=sharing). Please check the MILESTONES on the document, and let me know if there is something missing. 

So basically my current approach is to implement each identified function, in **go**. 
On this PR I have implemented the first 5 functions: `Info_error_exit, set_sysctl, set_host_networking_settings, add_to_path, setup_kind`

@jayunit100 , @rajaskakodkar please comments. Let me know if this is the way to go.

### CODE REVIEW

**Comment by jayunit100** [comment]
Thanks can you attach the cover sample output to a README.report.md file or something like that in this dir?

**Comment by emabota** [comment]
I don't understand why this is failing - https://github.com/kubernetes-sigs/kpng/actions/runs/4480600683/jobs/7876158548

**Comment by jayunit100** [comment]
https://github.com/kubernetes/kubernetes/issues/108891 <-- the failure is a flake.  just push a readme commit or something to rerun the testss

File: hack/test_e2e.go [hunk 1]
```diff
@@ -0,0 +1,944 @@
...
+	cmd = exec.Command("bash", "-c", cmd_string)
+	err = cmd.Run()
+	if err != nil {
+		log.Fatalf("Failed to create the file KUBECONFIG_TESTS.\n", err)
+	}
+
+	fmt.Printf("Cluster %s is created.\n", cluster_name)
+
+}
+
```
Comment (jayunit100) [inline hack/test_e2e.go:675]: godocccc

File: hack/test_e2e.go [hunk 2]
```diff
@@ -0,0 +1,944 @@
...
+		cmd = exec.Command("kubectl", "--context", k8s_context, "get", "pods", "--all-namespaces")
+		err = cmd.Run()
+		if err != nil {
+			log.Fatal("Error getting pods from all namespaces.", err)
+		}
+	}
+	fmt.Printf("%s is operational.\n",cluster_name)
+}
+
```
Comment (jayunit100) [inline hack/test_e2e.go:718]: add godoc

```
// install_kpng...
```

File: hack/test_e2e.go [hunk 3]
```diff
@@ -0,0 +1,944 @@
...
+	cmd = exec.Command(kubectl, "--context", k8s_context, "--namespace", NAMESPACE, "rollout", "status", "daemonset", "kpng", "-w", "--request-timeout", "3m")	
+	err = cmd.Run()
+	if err != nil {
+		log.Fatalf("Timeout waiting kpng rollout\n", err)
+	}
+
+	fmt.Println("Installation of kpng is done.")
+
+}
+
```
Comment (jayunit100) [inline hack/test_e2e.go:834]: godoc

File: hack/test_e2e.go [hunk 4]
```diff
@@ -0,0 +1,944 @@
+package main
+
```
Comment (jayunit100) [inline hack/test_e2e.go:2]: https://go.dev/blog/godoc

**Review by jayunit100 — CHANGES_REQUESTED** [review]

File: hack/test_e2e.go [hunk 5]
```diff
@@ -0,0 +1,944 @@
...
+		err = cmd.Run() 
+		if err != nil {
+			fmt.Println("The e2e tests currently support docker and podman as the container engine. Please install either of them")
+			os.Exit(1)
+		}
+	}
+	fmt.Println("Detected Container Engine:", CONTAINER_ENGINE)
+}
+
+func container_build(CONTAINER_FILE string, ci_mode bool) {
```
Comment (daman1807) [inline hack/test_e2e.go:411]: best to use `containerFile` and `ciMode` as parameters, following go conventions

https://go.dev/doc/effective_go#:~:text=Finally%2C%20the%20convention%20in%20Go%20is%20to%20use%20MixedCaps%20or%20mixedCaps%20rather%20than%20underscores%20to%20write%20multiword%20names.

File: hack/test_e2e.go [hunk 6]
```diff
@@ -0,0 +1,944 @@
...
+	cmd_string = ""
+	for _, s := range cmd_string_set {
+		cmd_string += s
+	}
+	cmd = exec.Command("bash", "-c", cmd_string)
+	err = cmd.Run()
+	if err != nil {
+		log.Fatalf("Cannot patch kube-proxy.\n", err)
+	}
+	fmt.Println("Kube-proxy patched! Guys how can I test this? To find out if it was really successful.")
```
Comment (daman1807) [inline hack/test_e2e.go:654]: "Guys how can I test this? To find out if it was really successful." 😅 
I guess if the `err` is nil, the cmd executed successfully.

File: hack/test_e2e.go [hunk 7]
```diff
@@ -0,0 +1,944 @@
...
+	if E2E_DEPLOYMENT_MODEL == "single-process-per-node" {
+		E2E_BACKEND_ARGS="'kube', '--kubeconfig=/var/lib/kpng/kubeconfig.conf', 'to-local', 'to-" + E2E_BACKEND + "', '--v=" + KPNG_DEBUG_LEVEL + "'"
+	} 
+
+	E2E_SERVER_ARGS = "[" + E2E_SERVER_ARGS + "]"
+	E2E_BACKEND_ARGS = "[" + E2E_BACKEND_ARGS + "]"
+
+	// Setting vars for generate the kpng deployment based on template
+	_ = os.Setenv("kpng_image", KPNG_IMAGE_TAG_NAME) 
+    _ = os.Setenv("image_pull_policy", "IfNotPresent")
```
Comment (daman1807) [inline hack/test_e2e.go:799]: identation

File: hack/test_e2e.go (hunk 4 above)
Comment (emabota) [reply hack/test_e2e.go:18]: Done

File: hack/test_e2e.go (hunk 5 above)
Comment (emabota) [reply hack/test_e2e.go]: Done

File: hack/test_e2e.go (hunk 1 above)
Comment (emabota) [reply hack/test_e2e.go:592]: Done

File: hack/test_e2e.go (hunk 2 above)
Comment (emabota) [reply hack/test_e2e.go:631]: Done

File: hack/test_e2e.go (hunk 7 above)
Comment (emabota) [reply hack/test_e2e.go:709]: @aroradaman I don't know why this is so. On my repo there is no identation issues  https://github.com/emabota/kpng/blob/8f0f3eb5229fee8eca3ff0650c5946d91d75582b/hack/test_e2e.go#L708
Any idea what's causing this?

File: hack/test_e2e.go (hunk 3 above)
Comment (emabota) [reply hack/test_e2e.go:744]: Done

File: hack/test_e2e.go [hunk 8]
```diff
@@ -0,0 +1,944 @@
...
+    // Add directory to path                                                   //
+    //                                                                         //
+    // Arguments:                                                              //
+    //   arg1:  directory                                                      //	
+	///////////////////////////////////////////////////////////////////////
+	_, err := os.Stat(directory)
+	if err != nil && os.IsNotExist(err) {
+		if_error_exit(err)
+	}
+
```
[the remark this answers is not in the document]
Comment (emabota) [reply hack/test_e2e.go:98]: Hi @aroradaman, after reading the link - [](https://pkg.go.dev/k8s.io/klog/v2) - I still don't understand why we should use logger instead of Println. At first I thought you were talking about the **Log** package! But then I realized that it's an all different package! Please can you help me understanding why **klog** is a better option? And wouldn't **Log** work here?

File: hack/test_e2e.go [hunk 9]
```diff
@@ -0,0 +1,944 @@
...
+    //   arg1: Path of dockerfile
+    //   arg2: ci_mode
+	///////////////////////////////////////////////////////////////////////////
+
+	detect_container_engine()
+	container_build(dockerfile, ci_mode) 
+
+}
+
+func create_cluster(cluster_name, bin_dir, ip_family, artifacts_directory string, ci_mode bool) {
```
[the remark this answers is not in the document]
Comment (emabota) [reply hack/test_e2e.go]: When kube-proxy is disabled the coredns pods FAIL to start! 
`$ kubectl get pods -n kube-system 
NAME                                                           READY   STATUS    RESTARTS   AGE
coredns-565d847f94-6chxd                                       0/1     Running   0          16m
coredns-565d847f94-86df5                                       0/1     Running   0          16m
etcd-kpng-e2e-ipv4-iptables-control-plane                      1/1     Running   0          16m
kindnet-f7fmc                                                  1/1     Running   0          15m
kindnet-fwtqn                                                  1/1     Running   0          16m
kindnet-n56n2                                                  1/1     Running   0          15m
kube-apiserver-kpng-e2e-ipv4-iptables-control-plane            1/1     Running   0          16m
kube-controller-manager-kpng-e2e-ipv4-iptables-control-plane   1/1     Running   0          16m
kube-scheduler-kpng-e2e-ipv4-iptables-control-plane            1/1     Running   0          16m
`

`$ kubectl get endpoints --all-namespaces 
NAMESPACE     NAME         ENDPOINTS         AGE
default       kubernetes   172.18.0.2:6443   16m
kube-system   kube-dns                       15m
`

[the remark this answers is not in the document]
Comment (emabota) [reply hack/test_e2e.go]:
```
$ kubectl get pods -n kube-system 
NAME                                                           READY   STATUS    RESTARTS   AGE
coredns-565d847f94-6chxd                                       0/1     Running   0          16m
coredns-565d847f94-86df5                                       0/1     Running   0          16m
etcd-kpng-e2e-ipv4-iptables-control-plane                      1/1     Running   0          16m
kindnet-f7fmc                                                  1/1     Running   0          15m
kindnet-fwtqn                                                  1/1     Running   0          16m
kindnet-n56n2                                                  1/1     Running   0          15m
kube-apiserver-kpng-e2e-ipv4-iptables-control-plane            1/1     Running   0          16m
kube-controller-manager-kpng-e2e-ipv4-iptables-control-plane   1/1     Running   0          16m
kube-scheduler-kpng-e2e-ipv4-iptables-control-plane            1/1     Running   0          16m
```

[the remark this answers is not in the document]
Comment (emabota) [reply hack/test_e2e.go]:
```
$ kubectl get endpoints --all-namespaces 
NAMESPACE     NAME         ENDPOINTS         AGE
default       kubernetes   172.18.0.2:6443   16m
kube-system   kube-dns                       15m

```

File: hack/test_e2e.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (aroradaman) [reply hack/test_e2e.go:98]: @emabota klog will give you the option for structured logging and you can control the verbosity ... also the logging will be consistent through kpng

File: hack/test_e2e.go [hunk 10]
```diff
@@ -0,0 +1,917 @@
+/*
+Test_e2e performs tests on the Kubernetes cluster. 
+It creates a kubernetes cluster based in kind + kpng, and performs ginkgo tests on it. 
+
+Usage:
```
Comment (knabben) [inline hack/test_e2e.go:5]: header license missing

File: hack/test_e2e.go [hunk 11]
```diff
@@ -0,0 +1,917 @@
...
+// commandExist check if a binary(cmdTest) exists. It returns True or False 
+func commandExist(cmdTest string) bool {
+	// Command is a Bash builtin command. 
+	// It execute a simple command or display information about commands.
+	// "command -v" print a description of COMMAND similar to the `type' builtin
+	cmdScript := "command -v " + cmdTest + " > /dev/null 2>&1"
+	cmd := exec.Command("bash", "-c", cmdScript)
+	err := cmd.Run()
+	
+	if err == nil {
```
Comment (knabben) [inline hack/test_e2e.go:86]: `return err == nil`

File: hack/test_e2e.go [hunk 12]
```diff
@@ -0,0 +1,917 @@
...
+	}
+
+	cmdString := "kubectl --context " + k8sContext + " wait --for=condition=ready pods --namespace=" + Namespace + " --selector k8s-app=kube-dns 1> /dev/null"
+	cmd := exec.Command("bash", "-c", cmdString)
+	err = cmd.Run()
+	if err != nil {
+		log.Fatal(err)
+	}
+	
+	if ciMode == true {
```
Comment (knabben) [inline hack/test_e2e.go:616]: `if ciMode {`

File: hack/test_e2e.go (hunk 2 above)
Comment (knabben) [reply hack/test_e2e.go:631]: please mark those godoc thread as resolved

File: hack/test_e2e.go (hunk 7 above)
Comment (knabben) [reply hack/test_e2e.go:709]: use `go fmt hack/test_e2e.go`

**Comment by rajaskakodkar** [comment]
@emabota just checking if you are still working on this

**Comment by emabota** [comment]
> @emabota just checking if you are still working on this

Hi @rajaskakodkar , yes, I am still interested. Just had a tight schedule lately, but have planned to resume it on June.
